federation/pkg/federation-controller/service/dns.go

```diff
@@ -263,6 +266,9 @@ func (s *ServiceController) ensureDnsRecords(clusterName string, cachedService *
 	serviceName := cachedService.lastState.Name
 	namespaceName := cachedService.lastState.Namespace
 	zoneNames, regionName, err := s.getClusterZoneNames(clusterName)
+	if zoneNames == nil {
+		return fmt.Errorf("fail to get cluster zone names")
```

Comment:
This should not be necessary, as catching err immediately below should cover this.  If not, there's a bug in getClusterZoneNames, as it should not be able to return err == nil, as well as zoneNames = nil.